src/subdivide_multi.h

```diff
@@ -0,0 +1,302 @@
+//
+//  main.cpp
+//  tet_subdivision
+//
+//  Created by Yiwen Ju on 12/2/23.
+//
+#pragma once
+
+//#define Non_Robust_Test
+//#define No_Multi_Check
+//#define Only_ZeroX
+//#define Only_Geometry
+//#define No_Multi_Check_3
+//#define Only_ZeroX_3
+//#define Only_Geometry_3
+//#define CSG_Base
+//#define MI_Base
+#include <iostream>
+#include <string>
+#include <valarray>
+#include <array>
+#include <SmallVector.h>
+#include <timer.h>
+#include <convex_hull_membership/contains.h>
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include <Eigen/Sparse>
+
+
+/// Two checks for each grid element: zero-crossing check and distance check. This function performs these two checks under the setting of implicit arrangement(IA), constructive solid geometry(CSG) and their curve networks.
+///
+/// @param[in] pts          4 arrays of three-tuples represent the coordinate of tet vertices. Each subarray represents a coordinate.
+/// @param[in] vals         functions' values at each vertex. Each subarray represents one function's values at four vertices.
+/// @param[in] grads            functions' gradients at each vertices. Each subarray represents one functions' gradients at four vertices and in each axis-aligned partial derivative.
+/// @param[in] threshold            The user-input epsilon value for the distance check.
+/// @param[out] active          A `bool` represents whether the tet is containing part of the geometry, i.e., this tet passes the zero-crossing test.
+///
+/// @return         A `bool` represents whether the tet is "refinable", i.e., passing the zero-crossing test and contains error greater than `threshold`.
+///
+///
+bool subTet(const std::array<std::array<double, 3>,4> &pts,
+            const llvm_vecsmall::SmallVector<std::array<double,4>, 20> &vals,
+            const llvm_vecsmall::SmallVector<std::array<std::array<double, 3>,4>, 20> &grads,
+            const double threshold,
+            bool& active);
```

Comment:
I would split this function, `subTet`, and the following `subMI` into three smaller functions for better clarity:
1. Build proxy
2. Zero cross check (for dim=2, 1, 0)
3. Distance check for (for dim=2, 1, 0)

Each function only do [one thing](https://cleancoders.com/episode/clean-code-episode-3). With these three smaller building block functions, the logic of IA, CSG, MI and Curve subdivision criteria should be much easier to implement.